### PR TITLE
Parallel coprocessor request

### PR DESCRIPTION
**Pull request by shenli**
@disksing @coocood 
I am not sure how to do unit tests for coprocessor client. So DNM. I will figure it out tomorrow.

### CODE REVIEW

File: store/tikv/coprocessor.go [hunk 1]
```diff
@@ -168,7 +291,7 @@ func (it *copIterator) Next() (io.ReadCloser, error) {
 		resp, err := client.SendCopReq(req)
 		if err != nil {
```
Comment (coocood) [inline store/tikv/coprocessor.go]: This error should be logged before `it.rebuildCurrentTask`, the error value in the following log would always be nil.

File: store/tikv/coprocessor.go [hunk 2]
```diff
@@ -142,19 +167,117 @@ type copIterator struct {
...
+					task = t
+					break
+				}
+			}
+		}
+		it.mu.Unlock()
+		if task == nil {
+			break
+		}
+		task.status = taskRunning
```
Comment (disksing) [inline store/tikv/coprocessor.go]: I think you should update the status with the mutex acquired, or 2 worker may send the same task.

Comment (coocood) [inline store/tikv/coprocessor.go]: task.status is concurrently accessed without lock.

File: store/tikv/coprocessor.go [hunk 3]
```diff
@@ -142,19 +167,117 @@ type copIterator struct {
...
+	errChan     chan error
+	finished    bool
 }
 
-func (it *copIterator) Next() (io.ReadCloser, error) {
-	// TODO: Support `Concurrent` instruction in `kv.Request`.
+// Pick the next new copTask and send request to tikv-server.
+func (it *copIterator) work() {
+	for {
+		if it.finished {
```
Comment (coocood) [inline store/tikv/coprocessor.go]: race access `it.finished`

File: store/tikv/coprocessor.go [hunk 4]
```diff
@@ -142,19 +167,117 @@ type copIterator struct {
...
+		}
+	} else {
+		var task *copTask
+		it.mu.Lock()
+		for _, t := range it.tasks {
+			if t.status != taskDone {
+				task = t
+				break
+			}
+		}
```
Comment (disksing) [inline store/tikv/coprocessor.go]: No need check `Desc`?

File: store/tikv/coprocessor.go [hunk 5]
```diff
@@ -202,36 +337,54 @@ func (it *copIterator) Next() (io.ReadCloser, error) {
...
 		}
-
-		if it.req.Desc {
-			it.index--
-		} else {
-			it.index++
-		}
-
-		return ioutil.NopCloser(bytes.NewBuffer(resp.Data)), nil
+		task.status = taskDone
```
Comment (coocood) [inline store/tikv/coprocessor.go]: need lock.

**Comment by coocood** [comment]
Reverse the task slice for descending order, the code would be simpler.

File: store/tikv/coprocessor.go [hunk 6]
```diff
@@ -142,19 +178,110 @@ type copIterator struct {
...
+	if err != nil {
+		it.Close()
+		return nil, err
+	}
+	it.mu.Lock()
+	defer it.mu.Unlock()
+	it.respGot++
+	if it.respGot == len(it.tasks) {
+		it.Close()
+	}
```
Comment (disksing) [inline store/tikv/coprocessor.go]: Is `respGot` redundant? Even without it, the iterator will be closed after all tasks' status are _Done_.

**Comment by disksing** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

**Comment by shenli** [comment]
PTAL @coocood @disksing

**Comment by coocood** [comment]
How about also add table scan time cost log?

**Comment by coocood** [comment]
LGTM

**Comment by disksing** [comment]
LGTM

**Pull request merged**